brave/browser/net/proxy/proxy_config_service_tor.cc

```diff
@@ -4,32 +4,141 @@
 
 #include "brave/browser/net/proxy/proxy_config_service_tor.h"
 
+#include <memory>
+#include <utility>
 #include <vector>
 
+#include "base/path_service.h"
 #include "base/strings/string_number_conversions.h"
 #include "base/strings/string_util.h"
+#include "chrome/common/chrome_paths.h"
 #include "crypto/random.h"
-#include "url/gurl.h"
+#include "net/proxy/proxy_service.h"
+#include "net/url_request/url_request_context.h"
+#include "services/service_manager/public/cpp/connector.h"
 
 namespace net {
 
+using content::BrowserThread;
 const int kTorPasswordLength = 16;
 
-ProxyConfigServiceTor::ProxyConfigServiceTor(const GURL& url)
-  : url_(url) {
-    config_.proxy_rules().ParseFromString(url_.spec());
+ProxyConfigServiceTor::ProxyConfigServiceTor(const std::string tor_path,
+  const std::string tor_proxy) {
+    if (tor_path.length() && tor_proxy.length()) {
+      url::Parsed url;
+      url::ParseStandardURL(
+        tor_proxy.c_str(),
+        std::min(tor_proxy.size(),
+                 static_cast<size_t>(std::numeric_limits<int>::max())),
+        &url);
+      if (url.scheme.is_valid()) {
+        scheme_ =
+          std::string(tor_proxy.begin() + url.scheme.begin,
+                      tor_proxy.begin() + url.scheme.begin + url.scheme.len);
+      }
+      if (url.host.is_valid()) {
+        host_ =
+          std::string(tor_proxy.begin() + url.host.begin,
+                      tor_proxy.begin() + url.host.begin + url.host.len);
+      }
+      if (url.port.is_valid()) {
+        port_ =
+          std::string(tor_proxy.begin() + url.port.begin,
+                      tor_proxy.begin() + url.port.begin + url.port.len);
+      }
+
+      base::WaitableEvent* tor_launched =
+        new base::WaitableEvent(
+          base::WaitableEvent::ResetPolicy::MANUAL,
+          base::WaitableEvent::InitialState::NOT_SIGNALED);
+      BrowserThread::PostTask(
+          BrowserThread::PROCESS_LAUNCHER, FROM_HERE,
+          base::Bind(&ProxyConfigServiceTor::LaunchTorProcess,
+                     base::Unretained(this),
+                     tor_launched,
+                     tor_path,
+                     host_,
```

Comment:
no need to pass `host_` and `port_` here. they are accessible in the member method

Reply:
and also save `tor_path` to `tor_path_` because you will use it in the case of `OnTorCrashed` which you missed